We will now consider the Boston housing data set, from the ISLP
library.

(a) Based on this data set, provide an estimate for the population mean of medv. Call this estimate μˆ.

In [1]:
from ISLP import load_data
import pandas as pd

boston = load_data('Boston')

mu_hat = boston['medv'].mean()
mu_hat

22.532806324110677

(b) Provide an estimate of the standard error of μˆ. Interpret this result.</br>

Hint: We can compute the standard error of the sample mean by</br>
dividing the sample standard deviation by the square root of the</br>
number of observations.</br>
計算樣本標準差：

$
s = \sqrt{\frac{1}{n-1} \sum_{i=1}^{n} (x_i - \hat{\mu})^2}
$

計算樣本均值的標準誤差：

$
SE(\hat{\mu}) = \frac{s}{\sqrt{n}}
$

In [2]:
import numpy as np

sample_std_dev = boston['medv'].std()

n = len(boston['medv'])


se_mu_hat = sample_std_dev / np.sqrt(n)
se_mu_hat


0.40886114749753505

假設標準誤差約為 0.409，這意味著我們的樣本均值在多次抽樣中，平均來說，大約會距離 真實母體均值0.409 單位。</br>
標準誤差越小，表示每次抽樣得到的樣本均值越接近母體均值。因此，如果標準誤差小，我們可以說 樣本均值是母體均值的更精確估計。</br>

(c) Now estimate the standard error of μˆ using the bootstrap. How does this compare to your answer from (b)?

In [5]:
import numpy as np

# 重抽樣次數
n_bootstraps = 1000
bootstrap_means = []

for _ in range(n_bootstraps):
    bootstrap_sample = boston['medv'].sample(frac=1, replace=True)
    bootstrap_means.append(bootstrap_sample.mean())

# 計算估計的標準誤
bootstrap_se_mu_hat = np.std(bootstrap_means)
bootstrap_se_mu_hat

0.4017780383924741

重抽樣過程會帶來一些額外的變動，每次隨機抽樣都會有點不確定性。</br>
兩者之間的差異不大，無論用哪種方法來估算，樣本均值都是相對穩定和可靠的，對母體均值的估計沒有太大影響。</br>

(d) Based on your bootstrap estimate from (c), provide a 95 % con-</br>
fidence interval for the mean of medv. Compare it to the results</br>
obtained by using Boston['medv'].std() and the two standard</br>
error rule (3.9).</br>
Hint: You can approximate a 95 % confidence interval using the
formula 
$
[\hat{\mu} - 2SE(\hat{\mu}), \hat{\mu} + 2SE(\hat{\mu})]
$


In [6]:
# Calculate the 95% confidence interval using the bootstrap standard error
ci_lower = mu_hat - 2 * bootstrap_se_mu_hat
ci_upper = mu_hat + 2 * bootstrap_se_mu_hat
(ci_lower, ci_upper)

(21.72925024732573, 23.336362400895624)

有 95% 的信心，真實的母體均值會落在這個範圍內(21.72925024732573, 23.336362400895624)

(e) Based on this data set, provide an estimate, μˆmed, for the median value of medv in the population.

In [7]:
# the median of medv as an estimate of the population median
mu_med_hat = boston['medv'].median()
mu_med_hat

21.2

(f) We now would like to estimate the standard error ofˆµmed. Unfortunately, there is no simple formula for computing the standard</br>
error of the median. Instead, estimate the standard error of the</br>
median using the bootstrap. Comment on your findings.</br>

In [15]:
# Set number of bootstrap samples
n_bootstraps = 1000
bootstrap_medians = []

# Perform bootstrap sampling
for _ in range(n_bootstraps):
    bootstrap_sample = boston['medv'].sample(frac=1, replace=True)
    bootstrap_medians.append(bootstrap_sample.median())

# Calculate the standard error of the median from bootstrap samples
bootstrap_se_mu_med = np.std(bootstrap_medians)
bootstrap_se_mu_med

0.3774992417210924

從母體中多次抽取樣本，計算每個樣本的中位數，這些樣本中位數大約會 偏離真實母體中位數 ± 0.377。
中位數的變動範圍 大約是 0.377

(g) Based on this data set, provide an estimate for the tenth per- centile of medv in Boston census tracts. </br>
Call this quantity μˆ0.1. (You can use the np.percentile() function.)</br>

In [16]:
# the 10th percentile of medv
mu_0_1_hat = np.percentile(boston['medv'], 10)
mu_0_1_hat

12.75

波士頓各區域的中位房價中，有 10% 的區域中位房價低於 12.75（單位：千美元）。
房價分布的低端情況，幫助我們了解哪些區域房價相對便宜。

(h) Use the bootstrap to estimate the standard error of μˆ0.1. Compercentile() ment on your findings.

In [17]:
n_bootstraps = 1000
bootstrap_percentiles = []

# Perform bootstrap sampling to estimate the standard error of the 10th percentile
for _ in range(n_bootstraps):
    bootstrap_sample = boston['medv'].sample(frac=1, replace=True)
    bootstrap_percentiles.append(np.percentile(bootstrap_sample, 10))

# Calculate the standard error of the 10th percentile from bootstrap samples
bootstrap_se_mu_0_1 = np.std(bootstrap_percentiles)
bootstrap_se_mu_0_1

0.5193538196643979

bootstrap 估算 波士頓房價資料集 medv 第 10 百分位數 的 標準誤差</br>
這個標準誤差 0.519 表示：如果我們重複從母體抽取樣本，每次計算第 10 百分位數，這些估計值 大約會在 ±0.519 的範圍內波動。</br>
第 10 百分位數是 12.75，但這個數字會因抽樣變異而略有上下浮動。</br>